tools/block-generator/generator/generate.go

```diff
@@ -375,100 +379,109 @@ func (g *generator) finishRound(txnCount uint64) {
 	g.pendingAssets = nil
 }
 
-// WriteBlock generates a block full of new transactions and writes it to the writer.
+// WriteBlock usually generates a block full of new transactions and writes it to the writer.
+// It also caches one round to allow the caller to request the same round multiple times.
+// This is motivated by the fact that Conduit's logic requests the initial round during
+// its Init() for catchup purposes, and once again when it starts ingesting blocks.
+// There are a few constraints on the generator arising from the fact that
+// blocks must be generated sequentially and that a fixed offset between the
+// database round and the generator round is presumed:
+//   - requested round < offset ---> error
+//   - requested round == offset: the generator will provide a genesis block or offset block
+//   - requested round == generator's round + offset ---> generate a block,
+//		advance the round, and cache the block in case of repeated requests.
+//   - requested round == generator's round + offset - 1 ---> write the cached block
+//		but do not advance the round.
+//   - requested round < generator's round + offset - 1 ---> error
 func (g *generator) WriteBlock(output io.Writer, round uint64) error {
 	if round < g.roundOffset {
 		return fmt.Errorf("cannot generate block for round %d, already in database", round)
 	}
-	if round-g.roundOffset != g.round {
-		return fmt.Errorf("generator only supports sequential block access. Expected %d but received request for %d", g.round+g.roundOffset, round)
-	}
-	numTxnForBlock := g.txnForRound(g.round)
+	virtualRound := g.round + g.roundOffset
+	if round+1 < virtualRound || virtualRound < round {
```

Comment:
I'm finding this expression very difficult to understand. I think this expresses the same thing with fewer conditions
```suggestion
	// return cached round if available
	if round - g.roundOffset + 1 == g.round && len(g.latestBlockMsgp) != 0{
		_, err := output.Write(g.latestBlockMsgp)
		if err != nil {
			return err
		}
	}
	// "genesis" round
	if round-g.roundOffset == 0 {
	}
	// previous logic
```

Reply:
Interestingly, for me, it's more difficult to grok this new approach. I guess each of our brains has different embedded neural nets for such inequalities.

Reply:
Trying to integrate this approach landed at a compromise, hopefully clearer. Introducing `nextRound` and `cachedRound` variables: 3c4fc1c354f14b392cd790faa956a0590e9a691c